content/en/blog/extend-falco-inputs-with-a-plugin-the-basics.md

```diff
@@ -0,0 +1,670 @@
+	}
+
+	err = json.Unmarshal(rawData, &data)
+	if err != nil {
+		fmt.Println(err.Error())
+		return err
+	}
+
+	switch req.Field() {
+	case "docker.status":
+		req.SetValue(data.Status)
+	case "docker.id":
+		req.SetValue(data.ID)
+	case "docker.from":
+		req.SetValue(data.From)
+	case "docker.type":
+		req.SetValue(data.Type)
+	case "docker.action":
+		req.SetValue(data.Action)
+	case "docker.scope":
+		req.SetValue(data.Scope)
+	case "docker.actor.id":
+		req.SetValue(data.Actor.ID)
+	case "docker.attributes.image":
+		req.SetValue(data.Actor.Attributes["image"])
+	case "docker.attributes.name":
+		req.SetValue(data.Actor.Attributes["name"])
+	case "docker.attributes.type":
+		req.SetValue(data.Actor.Attributes["type"])
+	default:
+		return fmt.Errorf("no known field: %s", req.Field())
+	}
+
+	return nil
+}
+
+// Open is called by Falco plugin framework for opening a stream of events, we call that an instance
+func (dockerPlugin *DockerPlugin) Open(params string) (source.Instance, error) {
+	return &DockerInstance{}, nil
+}
+
+// String represents the raw value of on event
+// (not currently used by Falco plugin framework, only there for future usage)
+func (dockerPlugin *DockerPlugin) String(in io.ReadSeeker) (string, error) {
+	evtBytes, err := ioutil.ReadAll(in)
+	if err != nil {
+		return "", err
+	}
+	evtStr := string(evtBytes)
+
+	return fmt.Sprintf("%v", evtStr), nil
+}
+
+// NextBatch is called by Falco plugin framework to get a batch of events from the instance
+func (dockerInstance *DockerInstance) NextBatch(pState sdk.PluginState, evts sdk.EventWriters) (int, error) {
+	dclient, err := docker.NewClientWithOpts()
+	if err != nil {
+		panic(err)
+	}
+
+	ctx := context.Background()
+	defer ctx.Done()
+
+	msg, _ := dclient.Events(ctx, dockerTypes.EventsOptions{})
+
+	e := [][]byte{}
+
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+
+	dockerPlugin := pState.(*DockerPlugin)
+
+L:
+	for {
+		expire := time.After(time.Duration(dockerPlugin.config.FlushInterval) * time.Second)
+		select {
+		case m := <-msg:
+			s, _ := json.Marshal(m)
+			e = append(e, s)
+			if len(e) >= evts.Len() {
+				break L
+			}
+		case <-expire:
+			if len(e) != 0 {
+				break L
+			}
+		case <-c:
+			return 0, sdk.ErrEOF
+		}
+	}
+
+	if len(e) == 0 {
+		return 0, nil
+	}
+
+	for n, i := range e {
+		evt := evts.Get(n)
+		_, err := evt.Writer().Write(i)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	return len(e), nil
+}
+
+// main is mandatory but empty, because the plugin will be used as C library by Falco plugin framework
+func main() {}
+```
+
+## Build
+
+The plugin is built as `c-shared` library, to get a `.so`:
+```shell
+go build -buildmode=c-shared -o /etc/falco/docker/libdocker.so
+```
+
+# Configuration
+
+Now we have our plugin, we must declare it to `Falco` in `falco.yaml`:
+
+```yaml
+plugins:
+  - name: docker
+    library_path: /etc/falco/docker/libdocker.so
+    init_config: '{"flush_interval": 1}'
```

Comment:
It failed:
```
Runtime error: Error reading config file(falco.yaml): could not load plugins config: yaml-cpp: error at line 6, column 7: bad conversion. Exiting.
```
So I prefer to not mention it :wink:

Reply:
@jasondellaluce any idea? :thinking:

Reply:
Do you think we can close this thread?

Reply:
Sorry for the delay. I think the problem here is that:
```
    ...
    init_config: 
      - flush_interval: 1
```
Should be turned into:
```
    ...
    init_config: 
      flush_interval: 1
```
That's because the `-` is defining a YAML sequence, right? So I have two thoughts about this:
- A sequence is not what you want in this case, I think
- I may add support to converting root-level sequences into JSON arrays into Falco

Reply:
Oh yeah, I had not noticed the `-` :man_facepalming:
@jasondellaluce you're right. 


@Issif I think this :point_down: is OK
```yaml
init_config: 
      flush_interval: 30
```

_Note that the interval is now expressed in milliseconds._